mat64/solve.go

```diff
@@ -0,0 +1,87 @@
+// Copyright ©2015 The gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package mat64
+
+import (
+	"github.com/gonum/blas"
+	"github.com/gonum/lapack/lapack64"
+)
+
+// Solve solves a minimum-norm solution to a system of linear equations defined
+// by the matrices a and b. If a is singular or near-singular a Condition error
+// is returned. Please see the documentation for Condition for more information.
+//
+// The minimization problem solved depends on the input parameters.
+//  1. If m >= n and trans == false, find x such that || a*x - b||_2 is minimized.
+//  2. If m < n and trans == false, find the minimum norm solution of a * x = b.
+//  3. If m >= n and trans == true, find the minimum norm solution of a^T * x = b.
+//  4. If m < n and trans == true, find X such that || a*x - b||_2 is minimized.
+// The solution matrix, x, is stored in place into the receiver.
+func (m *Dense) Solve(a, b Matrix) error {
+	ar, ac := a.Dims()
+	br, bc := b.Dims()
+	if ar != br {
+		panic(ErrShape)
+	}
+
+	aMat, _ := untranspose(a)
+	bMat, _ := untranspose(b)
+	m.reuseAs(ac, bc)
+	var restore func()
+	if m == aMat {
```

Comment:
> On Aug 15, 2015, at 10:58 PM, Dan Kortschak <notifications@github.com> wrote:
> 
> In mat64/solve.go <https://github.com/gonum/matrix/pull/199#discussion_r37142021>:
> 
> > +//  2. If m < n and trans == false, find the minimum norm solution of a * x = b.
> > +//  3. If m >= n and trans == true, find the minimum norm solution of a^T * x = b.
> > +//  4. If m < n and trans == true, find X such that || a*x - b||_2 is minimized.
> > +// The solution matrix, x, is stored in place into the receiver.
> > +func (m *Dense) Solve(a, b Matrix) error {
> > +	ar, ac := a.Dims()
> > +	br, bc := b.Dims()
> > +	if ar != br {
> > +		panic(ErrShape)
> > +	}
> > +
> > +	aMat, _ := untranspose(a)
> > +	bMat, _ := untranspose(b)
> > +	m.reuseAs(ac, bc)
> > +	var restore func()
> > +	if m == aMat {
> a.Copy(a.T()) was why I made it undefined. We have to allocate if we cover that. I comfortable with that decision. There needs to be a broader issue of shadowing. We can't detect all cases, but we can go a long way towards covering types that have a Raw* method.

It’s a reasonable decision for copy. I don’t know about the normal case. I’m not sure I want to have to reason about code like
c := a.View(1,2, 10, 9)
c.Solve(a,b)
d.Mul(a,b)